projects/github-actions/required-review/src/requirement.js

```diff
@@ -81,22 +85,42 @@ function buildReviewerFilter( config, teamConfig, indent ) {
 	if ( op === 'any-of' ) {
 		return async function ( reviewers ) {
 			core.info( `${ indent }Union of these:` );
-			return printSet( `${ indent }=>`, [
-				...new Set(
-					( await Promise.all( arg.map( f => f( reviewers, `${ indent }  ` ) ) ) ).flat( 1 )
-				),
-			] );
+			const reviewersAny = await Promise.all( arg.map( f => f( reviewers, `${ indent }  ` ) ) );
+			const requirementsMet = [];
+			const neededTeams = [];
+			for ( const requirementResult of reviewersAny ) {
+				if ( requirementResult.teamReviewers.length != 0 ) {
+					requirementsMet.push( requirementResult.teamReviewers );
+				};
+				if ( requirementResult.neededTeams.length != 0 ) {
+					neededTeams.push( requirementResult.neededTeams );
+				};
+			};
+			if( requirementsMet.some( a => a.length > 0 ) ){ // If there are requirements met, zero out the needed teams
+				neededTeams.length = 0;
+			};
+			return printSet( `${ indent }=>`, [ ...new Set( requirementsMet.flat( 1 ) )],  [ ...new Set( neededTeams.flat( 1 ) ) ]);
 		};
 	}
 
 	if ( op === 'all-of' ) {
 		return async function ( reviewers ) {
 			core.info( `${ indent }Union of these, if none are empty:` );
-			const filtered = await Promise.all( arg.map( f => f( reviewers, `${ indent }  ` ) ) );
-			if ( filtered.some( a => a.length === 0 ) ) {
-				return printSet( `${ indent }=>`, [] );
-			}
-			return printSet( `${ indent }=>`, [ ...new Set( filtered.flat( 1 ) ) ] );
+			const reviewersAll = await Promise.all( arg.map( f => f( reviewers, `${ indent }  ` ) ) );
+			const requirementsMet = [];
+			const neededTeams = [];
+			for ( const requirementResult of reviewersAll ) {
+				if ( requirementResult.teamReviewers.length != 0 ) {
+					requirementsMet.push( requirementResult.teamReviewers );
+				};
+				if ( requirementResult.neededTeams.length != 0 ) {
+					neededTeams.push( requirementResult.neededTeams );
+				};
+			};
+			if ( neededTeams.length !== 0 ) { // If there are needed teams, zero out requirements met
+				return printSet( `${ indent }=>`, [], [ ...new Set( neededTeams.flat( 1 ) ) ] );
+			};
+			return printSet( `${ indent }=>`, [ ...new Set( requirementsMet.flat( 1 ) ) ], [ ...new Set( neededTeams.flat( 1 ) ) ] );
```

Comment:
You know at this point that `neededTeams` is empty, so no need to run it through Set.
```suggestion
			return printSet( `${ indent }=>`, [ ...new Set( requirementsMet.flat( 1 ) ) ], [] );
```